daemon: Switch to new load-balancing control-plane

joamaki opened the pull request:
Enable the "enable-experimental-lb" by default to start using the new implementation. The experimental nomenclature will be removed in follow-up PRs and by v1.18 release the old control-plane will be removed.

The CiliumEnvoyConfig handling was updated to process listeners separately from the backends to properly handle multiple listeners referring to the same services.

```release-note
The service load-balancing control-plane in the Cilium agent has been redesigned which reduces memory usage and improves future extensibility of load-balancing features.
```

joamaki committed:
daemon: Switch to new load-balancing control-plane

Enable the "enable-experimental-lb" by default to start using the new
implementation. The experimental nomenclature will be removed in follow-up
PRs and by v1.18 release the old control-plane will be removed.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Update README

- Remove references to "experimental"
- Remove the comparisons to prior control-plane
- Add section on source code organization.
- Add notice about using 'stress.sh'.
- Add section on benchmarks.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Add stress.sh and watch.sh

Add the stress-test and watching scripts for working with the script tests.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/ciliumenvoyconfig/testdata/ingress.txtar:
```diff
@@ -460,3 +465,357 @@ SVC: ADDR=10.96.252.74:9080/TCP SLOT=0 LBALG=undef AFFTimeout=0 COUNT=1 QCOUNT=0
 SVC: ADDR=10.96.252.74:9080/TCP SLOT=1 BECOUNT=0 QCOUNT=0 FLAGS=ClusterIP+non-routable
 SVC: ADDR=10.96.44.54:9080/TCP SLOT=0 LBALG=undef AFFTimeout=0 COUNT=1 QCOUNT=0 FLAGS=ClusterIP+non-routable
```
TODO remove unused

mhofstetter commented on pkg/ciliumenvoyconfig/exp_tables.go:
```diff
@@ -125,15 +128,110 @@ func NewCECTable(db *statedb.DB) (statedb.RWTable[*CEC], error) {
+type EnvoyResourceListener struct {
+	Redirects           map[loadbalancer.ServiceName]*experimental.ProxyRedirect `json:"-"`
 	ReconciledRedirects map[loadbalancer.ServiceName]*experimental.ProxyRedirect `json:"-"`
+	ServicePorts        map[loadbalancer.ServiceName]sets.Set[string]
```
```
kind-worker> db/show envoy-resources --format=json
<stdin>:0: db/show envoy-resources: json.Marshal: json: unsupported type: map[loadbalancer.ServiceName]sets.Set[string]
```

:cry:

joamaki replied:
Oof. Really need to make it a habit to do tests for the marshalling (e.g. like in `pkg/loadbalancer/experimental/testdata/marshalling.txtar`). Will add the tests and fix this.

joamaki replied:
Added marshalling tests and actually changed from `map` to `part.Map` so I can properly marshal `loadbalancer.ServiceName` as a key.

mhofstetter commented on pkg/ciliumenvoyconfig/exp_reflector.go:
```diff
@@ -133,13 +136,36 @@ func registerCECReflector(
 				Namespace: objMeta.GetNamespace(),
 			},
 			Selector:         selector,
 			SelectsLocalNode: selectsLocalNode,
```
side-note: do i see this correctly that due to the use of `nodeLabels` there could be potentially a change of the `SelectsLocalNode` value at the startup of the agent - that potentially gets transferred to Envoy? (if the k8s-to-table reflector starts before the `controller` sets the actual node labels to the atomic map)

I know it's potentially only for short amount of time and eventually consistent. It's just that we had issues which such unnecessary updates in the past (as this might interrupt traffic)

joamaki replied:
Ah good catch. Yeah there is a possibility that it's not been filled yet. I added an "initialized" channel to `nodeLabels` which gets closed after its received the first time and it's checked on `Load()`. Still hoping we'll get the `Table[Node]` changes merged before v1.18 so we don't actually need to ship with this "hack"... :)

mhofstetter commented on pkg/ciliumenvoyconfig/exp_controller.go:
```diff
@@ -175,15 +176,77 @@ func (c *cecController) processLoop(ctx context.Context, health cell.Health) err
+			res = res.Clone()
+
+			// Look up the referenced service for the port name to port number mappings.
+			svc, _, watchSvc, found := c.Services.GetWatch(wtxn, experimental.ServiceByName(res.ClusterServiceName()))
```
Looks like the dependency `Backends statedb.Table[*experimental.Backend]` on `cecControllerParams` isn't used at all -> AFAIU  thery are retrieved via `Writer.BackendsForService` & a service reference from the service table

joamaki replied:
Ah indeed. Forgot to remove it when I removed the `bes Table[*Backend]` argument from `BackendsForService`.

mhofstetter commented on pkg/ciliumenvoyconfig/exp_controller.go:
```diff
@@ -140,7 +140,8 @@ func getProxyRedirect(cec *CEC, svcl *ciliumv2.ServiceListener) *experimental.Pr
 }
 
 func (c *cecController) processLoop(ctx context.Context, health cell.Health) error {
```
i know it's kind of out of scope for this PR - but IMO it would help to split the `cecController` into two separate controllers (structs & files) 

* ~ `nodeLabelController`
* ~ `cecController`

And add some godoc on top of these to describe what they are doing.

This should also help to reduce their dependencies

mhofstetter replied:
The `nodeLabelController` could also provide the `nodeLabels` (that is currently provided separately) - as it's actually updated and provided by this logic.

mhofstetter replied:
Note: (already mentioned offline) if we split this up - i'd suggest to use `job.Observer` instead of `job.OneShot` for nodelabelcontroller (`LocalNodeStore` is already an `Observable`)

joamaki replied:
Moved it into its own thing and using job.Observer now!

mhofstetter commented on pkg/ciliumenvoyconfig/exp_reflector.go:
```diff
@@ -6,6 +6,7 @@ package ciliumenvoyconfig
 import (
```
nit (i know also kind of scope for this PR): IMO it would help to rename this file to `exp_k8s_reflector.go` (& `registerCECK8sReflector`). AFAIU there could potentially be other reflectors than k8s too.

joamaki replied:
Makes sense! Renaming.

mhofstetter replied:
seems like this isn't renamed yet

joamaki replied:
Oopsie, renamed for real now :D

mhofstetter commented on pkg/ciliumenvoyconfig/exp_tables.go:
```diff
@@ -125,15 +128,110 @@ func NewCECTable(db *statedb.DB) (statedb.RWTable[*CEC], error) {
+}
+
+const (
+	EnvoyResourceKindListener  = EnvoyResourceKind("listener")
```
I understand the temptation to split this table into two separate "kinds" (listeners & endpoints). But this split can also create confusion. There can be cases (not for L7LB though) where a `CiliumEnvoyConfig` also contains Envoy endpoints. In these cases the CEC would be treated as `listeners` - without providing the properties that are special to that kind (referencedBy & listing of k8s services/IPs) 

Thinking out loud I wonder whether it would help to NOT separate between the two kinds and instead still create two entries - and prefix by their "origin" instead. And maybe even split this up into two controllers.

* controller 1: watches table `CEC` and simply creates entries in the table `EnvoyResource` (and keeps them in sync) - by checking the `SelectsLocalNode` - there name would be prefixed with `cec-` (instead of `listener:`)
* controller 2: watches table `CEC` and `Frontends` and simply creates entries in the table `EnvoyResource` for the backends that belong to frontends that are referenced by a CEC (that is selecting the local node) -> there name would be prefixed with `backendsync-`

(maybe we still kind of have to track the references and redirects internally - but i would prefer to keep the EnvoyResource table kind of generic without focussing too much on listeners & endpoints. Maybe even by listing the number (or even names) of all envoy resources that are part of table entry

* listeners
* routes
* clusters
* endpoints
* secrets

There's a high chance that i'm missing something (e.g. transactional aspects when splitting this up into multiple controllers and process loops) - maybe worth to quickly chat offline over my idea.

joamaki replied:
I'll rethink the naming a bit. Calling it "origin" rather than "kind" makes sense!

I would prefer to keep a single goroutine and processing loop so that we commit both the listeners and the cluster load assignments in a single write transaction. But I'll try to make it clear that the processing of CECs is conceptually separate from processing the backends by having the logic in separate methods.

mhofstetter reviewed (commented):
Thanks :rocket: I left some feedback and questions.

joamaki committed:
ciliumenvoyconfig: Process backends separately from listeners

As with the prior implementation, keep the "cec" resources separate from
the "backendsync" resources since there can be multiple listeners referring to the
same clusters ("services") and these should be able to be changed separately,
e.g. listeners shouldn't need to be recomputed if backends change.

Switch to using prototext for rendering the envoy resources so that we're
checking for the whole message. Since prototext is on purpose not deterministic
don't compare the marshalled format but instead compare the unmarshalled messages.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb reviewed (approved):
LGTM as sig-lb (:excited:)

mhofstetter commented on pkg/ciliumenvoyconfig/exp_controller.go:
```diff
@@ -42,24 +41,31 @@ type cecControllerParams struct {
 	Metrics        experimentalMetrics
 
 	NodeLabels     *nodeLabels
```
`LocalNodeStore`, `NodeLabels` & `Log` are unused.

joamaki replied:
Removed and double-checked all fields that they're used.

mhofstetter commented on pkg/ciliumenvoyconfig/exp_controller.go:
```diff
@@ -200,115 +160,234 @@ func (c *cecController) processLoop(ctx context.Context, health cell.Health) err
-func (c *cecController) processCEC(wtxn statedb.WriteTxn, cecName CECName) *statedb.WatchSet {
-	cec, _, watch, found := c.CECs.GetWatch(wtxn, CECByName(cecName))
+// cecProcessor computes desired envoy resources from the CECs. It upserts a EnvoyResource
+// with Origin=cec for the resouces coming from [CEC] and it upserts an EnvoyResource with
```
nit: typo `resources` -> `resources`

mhofstetter reviewed (commented):
Thanks for the changes! looks like we should remove the unused dependencies from the cec controller - then we're good to go :rocket: :tada:

joamaki committed:
gateway-api: Always generate a RouteConfiguration

As the http connection manager references the route configurations
we must generate one to avoid a timeout on Envoy side even if there
are no routes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
redirectpolicy: Fix flaky node-local-dns.txtar

By adding the pods in one go the order in which they are processed
is non-deterministic. Change the test to first add one pod and then
after it has been processed add the second pod.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -2218,7 +2218,7 @@ loadBalancer:
   # -- experimental enables support for the experimental load-balancing
   # control-plane.
-  experimental: false
+  experimental: true
```
This will not just enable this support in new clusters, but also enable it when upgrading existing clusters. 

```
$ helm status cilium -n kube-system
NAME: cilium
LAST DEPLOYED: Mon Apr  7 06:59:46 2025
NAMESPACE: kube-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
You have successfully installed Cilium with Hubble.

Your release version is 1.18.0-dev.

For any further help, visit https://docs.cilium.io/en/v1.18/gettinghelp
$ helm get values cilium -n kube-system --all | grep experimental
  experimental: false
$ helm upgrade cilium --dry-run -n kube-system --set upgradeCompatibility=1.17 oci://quay.io/cilium-charts-dev/cilium --version 1.18.0-dev-dev.272-f557d26c51 | grep experimental
Pulled: quay.io/cilium-charts-dev/cilium:1.18.0-dev-dev.272-f557d26c51
Digest: sha256:6174327bb9183ccefe39a2ef9f10f0156348089e880a02246ee799a0c830825a
  enable-experimental-lb: "true"
```

This is why we have the `upgradeCompatibility` flag, so that we can ensure that users on older releases don't get automatically migrated onto the new thing without realizing it. Is this what we want?

joamaki replied:
The flag will be removed very soon so won't be part of v1.18 release.
